### PR TITLE
Add more apis to Customer Insights swagger

### PR DESCRIPTION
**Pull request by oucfei**
This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 

### PR information
- [ ] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [ ] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes.
- [ ] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [ ] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [ ] My spec meets the review criteria:
  - [ ] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [ ] The spec follows the guidelines described in the [Swagger checklist](../documentation/swagger-checklist.md) document.
  - [ ] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

### CODE REVIEW

**Comment by jhendrixMSFT** [comment]
@oucfei Hello!  I'll be reviewing this today.  In the future could you please avoid moving existing APIs around in the doc?  The way the new operations have been added in this PR makes the diff unnecessarily complicated.

**Comment by oucfei** [comment]
@jhendrixMSFT yes... sorry about that, I was just trying to make the swagger more organized

File: specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-01-01/customer-insights.json [hunk 1]
```diff
@@ -203,7 +207,17 @@
...
             "required": true,
             "type": "string",
-            "description": "The name of the hub."
+            "description": "The name of the Hub.",
+            "pattern": "^[a-zA-Z][a-zA-Z0-9]+$",
+            "minLength": 1,
+            "maxLength": 64
+          },
+          {
+            "name": "organizationUrl",
```
Comment (jhendrixMSFT) [inline specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-01-01/customer-insights.json]: Where is this used?  There is a semantic validation error due to this, see the [log](https://travis-ci.org/Azure/azure-rest-api-specs/jobs/286291712).

**Review by jhendrixMSFT — CHANGES_REQUESTED** [review]

File: specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-01-01/customer-insights.json (hunk 1 above)
Comment (oucfei) [reply specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-01-01/customer-insights.json]: Fixed this. The "organizationUrl" should be in the query segment of the URL

**Comment by jhendrixMSFT** [comment]
@oucfei do you have a list of the added operations so I can review them outside of the diff?

**Comment by oucfei** [comment]
@jhendrixMSFT  Here's the list of added operations by operation id:
Hubs_SuggestTypeSchema
CrmMetadata_GetDiscoverorganizations
CrmMetadata_GetDiscoverentities
ConnectorMappings_Activate
Connectors_Activate
Connector_GetOperation
Segments_ListByHub
Segments_Get
Interaction_GetOperation
Interactions_Delete
Kpi_GetOperation
Link_GetOperation
PredictiveMatch_CreateOrUpdate
PredictiveMatch_Delete
PredictiveMatch_Get
PredictiveMatch_ListByHub
Profile_GetOperation

**Comment by jhendrixMSFT** [comment]
Thanks for the list.  It looks like these APIs were added to both API versions, is that correct?  Is this just a case that the swagger was missing them, meaning this functionality was already implemented for both API versions?

File: specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-04-26/customer-insights.json [hunk 2]
```diff
@@ -65,6 +65,449 @@
...
+          }
+        }
+      }
+    },
+	"/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.CustomerInsights/hubs/{hubName}/CrmMetadata/discoverorganizations": {
+      "get": {
+        "tags": [
+          "CrmMetadata"
+        ],
+        "operationId": "CrmMetadata_GetDiscoverorganizations",
```
Comment (jhendrixMSFT) [inline specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-04-26/customer-insights.json]: The operation name should be pascal cased, CrmMetadata_GetDiscoverOrganizations, and camel-cased in the path.

File: specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-01-01/customer-insights.json [hunk 3]
```diff
@@ -114,31 +114,27 @@
...
+      }
+    },
+	"/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.CustomerInsights/hubs/{hubName}/CrmMetadata/discoverorganizations": {
+      "post": {
         "tags": [
-          "Hubs"
+          "CrmMetadata"
         ],
-        "operationId": "Hubs_Update",
+        "operationId": "CrmMetadata_GetDiscoverorganizations",
```
Comment (jhendrixMSFT) [inline specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-01-01/customer-insights.json]: Pascal-case CrmMetadata_GetDiscoverOrganization.

File: specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-01-01/customer-insights.json [hunk 4]
```diff
@@ -172,24 +171,29 @@
...
+      }
+    },
+	"/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.CustomerInsights/hubs/{hubName}/CrmMetadata/discoverentities": {
+      "get": {
         "tags": [
-          "Hubs"
+          "CrmMetadata"
         ],
-        "operationId": "Hubs_Delete",
+        "operationId": "CrmMetadata_GetDiscoverentities",
```
Comment (jhendrixMSFT) [inline specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-01-01/customer-insights.json]: CrmMetadata_GetDiscoverEntities.

File: specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-04-26/customer-insights.json [hunk 5]
```diff
@@ -65,6 +65,449 @@
...
+          }
+        }
+      }
+    },
+	"/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.CustomerInsights/hubs/{hubName}/CrmMetadata/discoverentities": {
+      "get": {
+        "tags": [
+          "CrmMetadata"
+        ],
+        "operationId": "CrmMetadata_GetDiscoverentities",
```
Comment (jhendrixMSFT) [inline specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-04-26/customer-insights.json]: CrmMetadata_GetDiscoverEntities.

File: specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-04-26/customer-insights.json [hunk 6]
```diff
@@ -65,6 +65,449 @@
...
+          }
+        }
+      }
+    },
+	"/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.CustomerInsights/hubs/{hubName}/connectors/{connectorName}/activate": {
+      "post": {
+        "tags": [
+          "Connectors"
+        ],
+        "operationId": "Connectors_activate",
```
Comment (jhendrixMSFT) [inline specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-04-26/customer-insights.json]: Connectors_Activate.

File: specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-01-01/customer-insights.json [hunk 7]
```diff
@@ -260,26 +292,26 @@
...
-    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.CustomerInsights/hubs": {
-      "get": {
+	"/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.CustomerInsights/hubs/{hubName}/connectors/{connectorName}/activate": {
+      "post": {
         "tags": [
-          "Hubs"
+          "Connectors"
         ],
-        "operationId": "Hubs_ListByResourceGroup",
+        "operationId": "Connectors_activate",
```
Comment (jhendrixMSFT) [inline specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-01-01/customer-insights.json]: Connectors_Activate.

File: specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-01-01/customer-insights.json [hunk 8]
```diff
@@ -330,29 +404,26 @@
...
-    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.CustomerInsights/hubs/{hubName}/profiles/{profileName}": {
-      "put": {
+	"/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.CustomerInsights/hubs/{hubName}/segments/": {
+      "get": {
         "tags": [
-          "Profiles"
+          "segments"
         ],
-        "operationId": "Profiles_CreateOrUpdate",
+        "operationId": "segments_ListByHub",
```
Comment (jhendrixMSFT) [inline specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-01-01/customer-insights.json]: Segments_ListByHub.

File: specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-01-01/customer-insights.json [hunk 9]
```diff
@@ -396,28 +448,29 @@
...
+      }
+    },
+	"/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.CustomerInsights/hubs/{hubName}/segments/{segmentName}": {
       "get": {
         "tags": [
-          "Profiles"
+          "segments"
         ],
-        "operationId": "Profiles_Get",
+        "operationId": "segments_Get",
```
Comment (jhendrixMSFT) [inline specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-01-01/customer-insights.json]: Segments_Get.

File: specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-04-26/customer-insights.json [hunk 10]
```diff
@@ -65,6 +65,449 @@
...
+          }
+        }
+      }
+    },
+	"/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.CustomerInsights/hubs/{hubName}/segments/": {
+      "get": {
+        "tags": [
+          "segments"
+        ],
+        "operationId": "segments_ListByHub",
```
Comment (jhendrixMSFT) [inline specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-04-26/customer-insights.json]: Segments_ListByHub.

File: specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-04-26/customer-insights.json [hunk 11]
```diff
@@ -65,6 +65,449 @@
...
+          "nextLinkName": "nextLink"
+        }
+      }
+    },
+	"/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.CustomerInsights/hubs/{hubName}/segments/{segmentName}": {
+      "get": {
+        "tags": [
+          "segments"
+        ],
+        "operationId": "segments_Get",
```
Comment (jhendrixMSFT) [inline specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-04-26/customer-insights.json]: Segments_Get.

File: specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-04-26/customer-insights.json [hunk 12]
```diff
@@ -633,6 +1076,122 @@
...
+          }
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.CustomerInsights/hubs/{hubName}/interactions/{interactionName}/operations/{operation}": {
+      "get": {
+        "tags": [
+          "Interactions"
+        ],
+        "operationId": "Interaction_GetOperation",
```
Comment (jhendrixMSFT) [inline specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-04-26/customer-insights.json]: Interactions_GetOperation.

File: specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-01-01/customer-insights.json [hunk 13]
```diff
@@ -488,16 +535,822 @@
...
+          }
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.CustomerInsights/hubs/{hubName}/interactions/{interactionName}/operations/{operation}": {
+      "get": {
+        "tags": [
+          "Interactions"
+        ],
+        "operationId": "Interaction_GetOperation",
```
Comment (jhendrixMSFT) [inline specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-01-01/customer-insights.json]: Interactions_GetOperation.

**Review by jhendrixMSFT — CHANGES_REQUESTED** [review]

**Comment by oucfei** [comment]
@jhendrixMSFT fixed the above naming issues. For your question, yes those APIs were added to both API versions because the swagger was missing them. They were implemented in both versions already.

**Comment by oucfei** [comment]
@jhendrixMSFT @ravbhatnagar hi can you please finish reviewing this PR?

**Comment by jhendrixMSFT** [comment]
Ping @ravbhatnagar :)

File: specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-01-01/customer-insights.json [hunk 14]
```diff
@@ -330,29 +404,26 @@
         ],
         "responses": {
           "200": {
-            "description": "OK. Successfully get all the hubs in the subscription.",
+            "description": "OK. Successfully get the connector operation.",
             "schema": {
-              "$ref": "#/definitions/HubListResult"
+              "$ref": "#/definitions/ConnectorResourceFormat"
```
Comment (ravbhatnagar) [inline specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-01-01/customer-insights.json]: Why is GET on operation resource returning a connector resource. What is the intent of this API? What operation is this? RBAC issue.

File: specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-01-01/customer-insights.json [hunk 15]
```diff
@@ -330,29 +404,26 @@
...
           }
-        },
-        "x-ms-pageable": {
-          "nextLinkName": "nextLink"
         }
       }
     },
-    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.CustomerInsights/hubs/{hubName}/profiles/{profileName}": {
-      "put": {
+	"/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.CustomerInsights/hubs/{hubName}/segments/": {
```
Comment (ravbhatnagar) [inline specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-01-01/customer-insights.json]: Feedback about this API -
1. remove the / at the end of the API as this is a collection GET
2. provisioningState property needs to support Succeeded, Failed and Canceled terminal states.
3. What is the name proerpty inside properties. name is already there as a top level property outside properties.
4. Mark the read only properties appropriately. Ex - segmentId, created, version is most likely readonly.
5. rename created to creationTime and make the format as date-time to make is ISO8601 compliant.

File: specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-01-01/customer-insights.json [hunk 16]
```diff
@@ -457,24 +502,26 @@
...
-              "$ref": "#/definitions/ProfileResourceFormat"
+              "$ref": "#/definitions/SegmentResourceFormat"
             }
           }
         }
-      },
-      "delete": {
+      }
+	},
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.CustomerInsights/hubs/{hubName}": {
```
Comment (ravbhatnagar) [inline specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-01-01/customer-insights.json]: Feedback about this API -
hubBillingInfo - why are you not using the Sku property for this as defined in the ARM RPC.
Mark read only properties appropriately. Add allowed values for provisioningState property

File: specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-01-01/customer-insights.json [hunk 17]
```diff
@@ -488,16 +535,822 @@
...
+            "description": "Parameters supplied to the CreateOrUpdate Hub operation."
+          },
+          {
+            "$ref": "#/parameters/ApiVersionParameter"
+          },
+          {
+            "$ref": "#/parameters/SubscriptionIdParameter"
+          }
+        ],
+        "responses": {
```
Comment (ravbhatnagar) [inline specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-01-01/customer-insights.json]: Model a "default" response for all the APIs which captures the error cases and the response is compliant with ARM RPC error contract. Example here - https://github.com/Azure/azure-rest-api-specs/blob/current/specification/batch/resource-manager/Microsoft.Batch/2017-05-01/BatchManagement.json#L149

**Review by ravbhatnagar — COMMENTED** [review]
@oucfei I have provided some feedback and there are some basic/important API modeling questions which apply across the changes in the review. So I am not completing my review. I recommend schedule a meeting to close this review in person.

File: specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-01-01/customer-insights.json [hunk 18]
```diff
@@ -114,31 +114,27 @@
...
-              "$ref": "#/definitions/Hub"
+			  "$ref": "#/definitions/SuggestTypeSchemaActionResult"
             }
           }
         }
-      },
-      "patch": {
+      }
+    },
+	"/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.CustomerInsights/hubs/{hubName}/CrmMetadata/discoverorganizations": {
```
[the remark this answers is not in the document]
Comment (oucfei) [reply specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-01-01/customer-insights.json]: CrmMetadata is not a resource type... I don't know why this API is designed this way, it's not very RESTful, but it is indeed looks like this, and we're not going to change this API at this time

File: specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-01-01/customer-insights.json [hunk 19]
```diff
@@ -172,24 +171,29 @@
...
+				"$ref": "#/definitions/CrmOrganization"
+			  }
             }
           }
         }
-      },
-      "delete": {
+      }
+    },
+	"/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.CustomerInsights/hubs/{hubName}/CrmMetadata/discoverentities": {
```
[the remark this answers is not in the document]
Comment (oucfei) [reply specification/customer-insights/resource-manager/Microsoft.CustomerInsights/2017-01-01/customer-insights.json]: same as above.

**Comment by ravbhatnagar** [comment]
@jhendrixMSFT  @salameer as FYI. I have asked @oucfei to schedule a sync with ARM. The APIs are not modeled correctly.

**Comment by jhendrixMSFT** [comment]
@oucfei @ravbhatnagar did you two meet to discuss the changes?

**Comment by oucfei** [comment]
@jhendrixMSFT yes, @ravbhatnagar explained the API design issues with our team. We decided to fix the RBAC related issues first, before checking in this PR.

**Comment by jhendrixMSFT** [comment]
@oucfei Thanks for the info, so does that mean that this PR will be updated once the RBAC changes have been made?

**Comment by oucfei** [comment]
@jhendrixMSFT yes I'll update the PR once the fix is checked in

**Comment by jhendrixMSFT** [comment]
@oucfei I'm going to close this while you're working on the back-end changes.  Once you're ready you can re-open this PR and update it with the changes.

**Pull request closed without merging**